### PR TITLE
Selected tag is not shown when isMutliple is false

### LINKED ISSUE #195: Selected tag is not shown when isMutliple is false
I have no clue if I am using it wrong, but if `->setIsMultiple(false)` is set a selected tag title is not displayed.
But the record in the many_many junction table is created and the `X` also shows that some relation is present.

Screenshot

### PRs
- https://github.com/silverstripe/silverstripe-tagfield/pull/201

### PR DESCRIPTION
**Pull request by wilr**
Fixes #195, #153

### CODE REVIEW

File: client/src/components/TagField.js [hunk 1]
```diff
@@ -128,12 +123,28 @@ class TagField extends Component {
...
+            Object.keys(passThroughAttributes.value)[0]
+          ];
+
+        if (typeof value === "object") {
+          passThroughAttributes.value = value;
+        }
+      }
+    }
+
+    console.log(this.props);
```
Comment (emteknetnz) [inline client/src/components/TagField.js]:
```suggestion
```

File: src/TagField.php [hunk 2]
```diff
@@ -304,36 +319,10 @@ protected function getOptions($onlySelected = false)
...
 
         $source->each($addOption);
+
         return $options;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function setValue($value, $source = null)
```
Comment (emteknetnz) [inline src/TagField.php:313]: This is backwards incompatible API change, we cannot remove this public function

File: src/TagField.php [hunk 3]
```diff
@@ -368,80 +358,125 @@ public function setSource($source)
...
+
+        if (is_array($values)) {
+            return $values;
+        }
+
+        if ($values instanceof SS_List) {
+            return $values->column($this->getTitleField());
+        }
+
+        return [trim($values)];
```
Comment (emteknetnz) [inline src/TagField.php]:
```suggestion
        return [trim((string) $values)];
```
?

File: src/TagField.php [hunk 4]
```diff
@@ -368,80 +358,125 @@ public function setSource($source)
...
-                'name' => $this->getName() . '[]',
+                'name' => $name,
                 'style' => 'width: 100%',
                 'data-schema' => json_encode($this->getSchemaData()),
             ]
         );
     }
 
+
+    protected function getListValues($values)
```
Comment (emteknetnz) [inline src/TagField.php]:
```suggestion
    protected function getListValues($values): array
```

File: client/src/components/TagField.js [hunk 5]
```diff
@@ -1,10 +1,10 @@
-import React, { Component } from 'react';
-import Select from 'react-select';
-import fetch from 'isomorphic-fetch';
-import fieldHolder from 'components/FieldHolder/FieldHolder';
-import url from 'url';
-import debounce from 'debounce-promise';
-import PropTypes from 'prop-types';
+import React, { Component } from "react";
```
Comment (emteknetnz) [inline client/src/components/TagField.js]: There's a lot of JS changes that look like your IDE may have done some automated formatting.  Double quotes are not standard in this codebase, so these change will need to be reverted. It's causing linting errors in travis - https://app.travis-ci.com/github/silverstripe/silverstripe-tagfield/jobs/552813829

File: readme.md [hunk 6]
```diff
@@ -2,23 +2,27 @@
 
 Custom tag input field, for SilverStripe.
 
-[![Build Status](https://api.travis-ci.com/silverstripe/silverstripe-tagfield.svg?branch=2)](https://travis-ci.com/silverstripe/silverstripe-tagfield)
-[![Code Quality](http://img.shields.io/scrutinizer/g/silverstripe/silverstripe-tagfield.svg?style=flat)](https://scrutinizer-ci.com/g/silverstripe/silverstripe-tagfield)
-[![Code coverage](https://codecov.io/gh/silverstripe/silverstripe-tagfield/branch/master/graph/badge.svg)](https://codecov.io/gh/silverstripe/silverstripe-tagfield)
+[![Build
```
Comment (emteknetnz) [inline readme.md]: These long lines weren't failing any existing linting rules, and now won't match other modules. Please revert

File: src/TagField.php [hunk 7]
```diff
@@ -567,4 +600,14 @@ public function getSchemaStateDefaults()
 
         return $data;
     }
+
+
+    public function getSchemaDataType()
```
Comment (emteknetnz) [inline src/TagField.php]:
```suggestion
    public function getSchemaDataType(): string
```

File: src/TagField.php [hunk 8]
```diff
@@ -368,80 +358,125 @@ public function setSource($source)
...
-     * @param  string $term
+     * @param string $value
+     *
      * @return DataObject|bool
      */
-    protected function getOrCreateTag($term)
+    protected function getOrCreateTag($value)
     {
+        if (is_array($value)) {
+            $value = isset($value['Value']) ? $value['Value'] : '';
```
Comment (emteknetnz) [inline src/TagField.php]:
```suggestion
            $value = $value['Value'] ?? '';
```

File: src/TagField.php [hunk 9]
```diff
@@ -567,4 +600,14 @@ public function getSchemaStateDefaults()
...
     }
+
+
+    public function getSchemaDataType()
+    {
+        if ($this->getIsMultiple()) {
+            return self::SCHEMA_DATA_TYPE_MULTISELECT;
+        } else {
+            return self::SCHEMA_DATA_TYPE_SINGLESELECT;
+        }
```
Comment (emteknetnz) [inline src/TagField.php]:
```suggestion
        if ($this->getIsMultiple()) {
            return self::SCHEMA_DATA_TYPE_MULTISELECT;
        }
        return self::SCHEMA_DATA_TYPE_SINGLESELECT;
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Looks good so far, I've left feedback and there's also a travis unit test failure to sort out too. I haven't manually replicated issue or tested PR yet.

File: src/TagField.php (hunk 2 above)
Comment (wilr) [reply src/TagField.php:313]: @emteknetnz it's not removing it from the API, it's defined on the parent as public anyway.

File: client/src/components/TagField.js (hunk 5 above)
Comment (wilr) [reply client/src/components/TagField.js]: Ok, single quotes wasn't configured in the `.editorconfig` I've added that preference now.

File: .editorconfig [hunk 10]
```diff
@@ -13,6 +13,7 @@ trim_trailing_whitespace = true
 [{*.yml,package.json,*.js,*.scss}]
 indent_size = 2
 indent_style = space
+quote_type = single
```
Comment (wilr) [inline .editorconfig:16]: @emteknetnz Added the preference for single quotes

**Comment by wilr** [comment]
@emteknetnz reverted readme change and rebased.

File: src/TagField.php (hunk 2 above)
Comment (emteknetnz) [reply src/TagField.php:313]: So why are we removing it here?  Would the parent method handle different inputs of $value and $source correctly?

Parent method `MultiSelectField::setValue()` is this for reference:
```php
    public function setValue($value, $obj = null)
    {
        // If we're not passed a value directly,
        // we can look for it in a relation method on the object passed as a second arg
        if ($obj instanceof DataObject) {
            $this->loadFrom($obj);
        } else {
            parent::setValue($value);
        }
        return $this;
    }
```

File: tests/StringTagFieldTest.php [hunk 11]
```diff
@@ -171,6 +171,33 @@ public function testSchemaIsAddedToAttributes()
         $this->assertNotEmpty($attributes['data-schema']);
     }
 
+    /**
+     * @param array $expected
+     * @param TagFieldTestBlogPost $record
+     */
+    protected function compareExpectedAndActualTags(array $expected, TagFieldTestBlogPost $record)
```
Comment (emteknetnz) [inline tests/StringTagFieldTest.php]: Why is method here?  It's not called within StringTagFieldTest?

**Review by emteknetnz — CHANGES_REQUESTED** [review]
This does not seem to fix the original issue?  Setting `->isMultiple(false)` for both `TagField` and `StringTagfield` meant the tags did not show in the field

Code I used for testing below:

```php
<?php
use SilverStripe\TagField\TagField;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\TagField\StringTagField;

class BlogPost extends SiteTree
{
    private static $many_many = [
        'BlogTags' => BlogTag::class
    ];

    private static $db = [
        'StringTags' => 'Text',
    ];

    public function getCMSFields()
    {
        $fields = parent::getCMSFields();
        /** @var TagField field */
        $field = TagField::create(
            'BlogTags',
            'Blog Tags',
            BlogTag::get(),
            $this->BlogTags()
        )
            ->setShouldLazyLoad(true)
            ->setCanCreate(true)
            ->setIsMultiple(true)
            ;

        $fields->insertAfter('Title', $field);

        $field = StringTagField::create(
            'StringTags',
            'StringTags',
            ['one', 'two'],
            explode(',', $this->Tags)
        )
            ->setShouldLazyLoad(true)
            ->setIsMultiple(true)
            ;
        $fields->insertAfter('Title', $field);

        return $fields;
    }
}

<?php

use SilverStripe\ORM\DataObject;

class BlogTag extends DataObject
{
    private static $db = [
        'Title' => 'Varchar(200)',
    ];

    private static $belongs_many_many = [
        'BlogPosts' => BlogPost::class
    ];
}

<?php

use SilverStripe\Admin\ModelAdmin;

class BlogTagAdmin extends ModelAdmin
{
    private static $url_segment = 'blog-tags';

    private static $managed_models = [
        BlogTag::class
    ];
}
```

**Comment by wilr** [comment]
@emteknetnz sorry when rebasing I didn't rebuild the bundle.js so the bug would still be present. If you try that on your checkout it should be green. Just checking it out now.

File: tests/StringTagFieldTest.php (hunk 11 above)
Comment (wilr) [reply tests/StringTagFieldTest.php]: Removed method.

**Comment by wilr** [comment]
@emteknetnz Updated the `bundle.js` correctly now after my re-base and confirmed that example works as intended (admin:1.10.2)

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Thanks, have tested locally, now behaves as expected

There's a failing unit test that will need to be resolved

**Comment by wilr** [comment]
Failing test in Postgres was because of the fallback to `ID`. It didn't like doing `ID IN ("Test1")`.

I've removed the ID fallback now.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
That last commit broken functionality - I manually tested that regular `TagField` no longer saves correctly when adding `->setIsMultiple(true)`.  Reverting the last commit fixes this issue, though obviously it means the postgres unit test will also fail.

**Comment by wilr** [comment]
@emteknetnz fixed that regression now.

**Review by emteknetnz — APPROVED** [review]
Tested locally, works good

**Pull request merged**